icu4j/main/classes/core/src/com/ibm/icu/message2/Mf2DataModelFormatter.java

```diff
@@ -0,0 +1,277 @@
+                if (selectors.size() != keysToCheck.size()) {
+                    throw new IllegalArgumentException("Mismatch between the number of selectors and the number of keys: "
+                            + selectors.size() + " vs. " + keysToCheck.size());
+                }
+                boolean matches = true;
+                // Iterate "horizontally", through all matching functions and keys
+                for (int i = 0; i < maxCount; i++) {
+                    Expression selector = selectors.get(i);
+                    String valToCheck = keysToCheck.get(i);
+                    Selector func = selectorFunctions.get(i);
+                    Map<String, Object> options = mf2OptToFixedOptions(selector.getOptions());
+                    if (!func.matches(variableToObjectEx(selector.getOperand(), arguments), valToCheck, options)) {
+                        matches = false;
+                        break;
+                    }
+                }
+                if (matches) {
+                    patternToRender = variant.getValue();
+                    break;
+                }
+            }
+        }
+
+        // TODO: check that there was an entry with all the keys set to `*`
+        if (patternToRender == null) {
+            // If there was a case with all entries in the keys `*` this should not happen
+            throw new IllegalArgumentException("The selection went wrong, cannot select any option.");
+        }
+
+        StringBuilder result = new StringBuilder();
+        for (Part part : patternToRender.getParts()) {
+            if (part instanceof Text) {
+                result.append(part);
+            } else if (part instanceof Expression) { // Placeholder is an Expression
+                FormattedPlaceholder fr = formatPlaceholder((Expression) part, arguments, false);
+                FormattedValue fv = fr.getFormattedValue();
+                if (fr.getInput() == null && (fv == null || fv.toString() == null)) {
+                    result.append("{" + ((Expression) part).getOperand() + "}");
+                } else {
+                    result.append(fr.toString());
+                }
+            } else {
+                throw new IllegalArgumentException("Unknown part type: " + part);
+            }
+        }
+        return result.toString();
+    }
+
+    private Object variableToObjectEx(Value value, Map<String, Object> arguments) {
+        return variableToObjectEx(value, arguments, 0);
+    }
+
+    /*
+     * Pass a level to prevent local variables calling each-other recursively:
+     *
+     * <code><pre>
+     * let $l1 = {$l4 :number}
+     * let $l2 = {$l1 :number}
+     * let $l3 = {$l2 :number}
+     * let $l4 = {$l3 :number}
+     * </pre></code>
+     *
+     * We can keep track of the calls (complicated and expensive).
+     * Or we can forbid the use of variables before they are declared, but that is not in the spec (yet?).
```

Comment:
+1 "KISS"

Reply:
ACK. I'm all for it.
We have an issue opened for MFWG to put that in the spec.